[FIX] survey: partner access to survey input

nle-odoo opened the pull request:
Allow the partner associated to a survey user input to read the data on
this output.

A survey can be created by a different user than the one
filling it, for example an interview request if the interviewer has not
created the survey, he can't see the answers or open the survey.

opw-644791

beledouxdenis commented:
This is surely associated to:
https://github.com/odoo/odoo/commit/812318dcbae13aa6ab87b919b31bc78cc74b97fe

In your use case, this isn't an invitation / token ?

nle-odoo commented:
it is a browse on user_input to have the token for launching an action. So yes and seeing your fix I could also use SUPERUSER_ID.

beledouxdenis commented:
Well, it's mostly the fact the survey can be filled without being connected / signed in.

So adding `('partner_id', '=', user.partner_id.id)` won't change anything.

nle-odoo commented:
I edited to show the part in the code:

`addons/hr_evaluation/hr_evaluation.py`

```py

    def action_print_survey(self, cr, uid, ids, context=None):                                      
        """ If response is available then print this response otherwise print survey form (print template of the survey) """
        context = dict(context or {})                                                               
        interview = self.browse(cr, uid, ids, context=context)[0]                                   
        survey_obj = self.pool.get('survey.survey')                                                 
        response_obj = self.pool.get('survey.user_input')                                           
        response = response_obj.browse(cr, uid, interview.request_id.id, context=context)           
        context.update({'survey_token': response.token})                                            
        return survey_obj.action_print_survey(cr, uid, [interview.survey_id.id], context=context)   
                                                                                                    
    def action_start_survey(self, cr, uid, ids, context=None):                                      
        context = dict(context or {})                                                               
        interview = self.browse(cr, uid, ids, context=context)[0]                                   
        survey_obj = self.pool.get('survey.survey')                                                 
        response_obj = self.pool.get('survey.user_input')                                           
        # grab the token of the response and start surveying                                        
        response = response_obj.browse(cr, uid, interview.request_id.id, context=context)           
        context.update({'survey_token': response.token})                                            
        return survey_obj.action_start_survey(cr, uid, [interview.survey_id.id], context=context)   
```

beledouxdenis commented:
? :D

nle-odoo commented:
i edited the previous comment when you answered so I added a comment to show my edit.

In my use case it was an action from the backend of a connected interviewer, so he was connected.

I replaced commit https://github.com/odoo-dev/odoo/commit/00e4e1df5ec5d4a06d05bede1567eba57ed348b5 with https://github.com/odoo-dev/odoo/commit/29ed74f619d39e9a45b1a48ca262443b54dc419d.

beledouxdenis commented:
Well, in my case, using SUPERUSER_ID was acceptable because the token provided was correct, granting him some kind of access.

In your case, calling `action_print_survey` with an xmlrpc call and any random id would allow you to print, and therefore access, data that might not be yours.

You said that in your case, the user was connected. From where did he prints his reports ? The portal ? Or this is in the website ?

nle-odoo commented:
In the backend, from Human Resources > Appraisal | Interview Requests > {go to one record} : 

![](http://i.imgur.com/ekTZpba.png)

but it happens when printing, and also when getting the link to answer the survey.

beledouxdenis commented:
I don't understand: You have access to these buttons without being (yet) an employee ?

nle-odoo commented:
It is for the module hr_evaluation, so it is to review the job quality of an existing employee.

So the employee which is evaluated is already an employee, and the Interviewer is the res.partner one who see these buttons.

In this use case:
* an employee is associated to the user,
* member of survey / user,
* employee level permission for Human Resource,
* the appraisal interview is created by another user,

and the employee evaluated and the interviewer are the same, but the issue also happens whenever:

* the user having created the interview request is not the same as the interviewer,
* AND the interviewer is not in the group Survey / Manager.

beledouxdenis commented:
I am completely confused.

This means that both users are employees (Both are part of the Human Resources > Employee) ?

I guess your first fix was better then. The fact I used SUPERUSER_ID in the related issue is because, in this case, the survey could be filled while not even being connected. This is not your case.
